sdk/dotnet/Pulumi/Deployment/Deployment.Runner.cs

```diff
@@ -29,9 +30,12 @@ private class Runner : IRunner
             /// continuously, asynchronously loop, waiting for these tasks to complete, and only
             /// exiting once the set becomes empty.
             /// </summary>
-            private readonly Dictionary<Task, List<string>> _inFlightTasks = new Dictionary<Task, List<string>>();
+            private readonly TaskMonitoringHelper _inFlightTasks = new TaskMonitoringHelper();
             private readonly List<Exception> _exceptions = new List<Exception>();
 
+            private readonly ConcurrentDictionary<Tuple<int,string>,int> _descriptions =
```

Comment:
you can use named tuples for this so it's a bit easier to read (though I will admit not easier to write).

Those look like:
```csharp
private readonly ConcurrentDictionary<(int NameOne, string NameTwo), int> _descriptions = ...;
```

Which can be expanded when you are using them like:
```csharp
var (nameOne, nameTwo) = this._descriptions.Keys[0];
```

Reply:
This is a thinking man's PR so i'll try to grok the rest of it over my coffee tmr lol